### ISSUE TITLE
stashApply succeeds but compilation still triggers

### ISSUE THREAD

**Issue by tyrcho** [romanowski/hoarder#70]
On all sbt projects at work we have this issue. 
We run `sbt test:compile stash` in Gitlab CI job and share the stashed folder as an artifact.
On the next job we run `sbt stashApply test`.
The compilation is still triggered before running the tests.

I've tried on this [open-source, simpler project](https://gitlab.com/tyrcho/approvals-scala) and I reproduce the issue.

I've added `addSbtPlugin("com.github.romanowski" % "hoarder" % "1.0.5")` in  `project\plugins.sbt` then I run these commands:

    sbt test:compile stash

I can see in the log that this succeeded: 

`[info] Project approvals-demo-scala stashed to /Users/michel/.sbt/1.0/sbt-stash/approvals-demo-scala/HEAD/2.12 using classes from List(ExportedCache(/Users/michel/.sbt/1.0/sbt-stash/approvals-demo-scala/HEAD/2.12/approvals-demo-scala/compile/analysis.zip,None), ExportedCache(/Users/michel/.sbt/1.0/sbt-stash/approvals-demo-scala/HEAD/2.12/approvals-demo-scala/test/analysis.zip,None))
`

    sbt clean 
    sbt stashApply test  

Here is the log where you can see that the compilation triggers even after stashApply:

```
[info] Set current project to approvals-demo-scala (in build file:/Users/michel/git/katas/approvals-demo-scala/)
[success] Total time: 0 s, completed Dec 26, 2019, 4:55:32 PM
[info] Stash for approvals-demo-scala applied from /Users/michel/.sbt/1.0/sbt-stash/approvals-demo-scala/HEAD/2.12 to List(/Users/michel/git/katas/approvals-demo-scala/target/scala-2.12/classes, /Users/michel/git/katas/approvals-demo-scala/target/scala-2.12/test-classes)
[success] Total time: 1 s, completed Dec 26, 2019, 4:55:33 PM
[info] Compiling 1 Scala source to /Users/michel/git/katas/approvals-demo-scala/target/scala-2.12/classes ...
[info] Compiling 4 Scala sources to /Users/michel/git/katas/approvals-demo-scala/target/scala-2.12/test-classes ...
```
    

I've checked the logs for pipelines since months and it seems this issue has been present for a long time, I have not found a CI job where the compilation was not triggered.

**Comment by romanowski** [comment]
Hi @tyrcho 

I am looking at the problem. So to seems that [this commit](https://github.com/sbt/sbt/commit/25e97f99f588d1ddc9bbd114740bfb25c2a955c6#diff-324c154efade6a91675ce59c16db4b9d) and follows ups by @eatkins added to sbt a custom mechanism to track changed sources and binaries. 

I need to investigate if I can plug in into its current form or I will need to create PR to sbt to support it.

cc @eed3si9n

**Comment by tyrcho** [comment]
Hi, thanks for replying. If I understand correctly, a recent change in sbt prevents it to pickup stashed files from hoarder and compilation is triggered again ?

**Comment by romanowski** [comment]
Yes. I am working on workaround (but I am still not sure if it will just works).

**Comment by rtoomey-coatue** [comment]
Thanks, I see this issue as well - using sbt 1.3.6 on AWS Codebuild (backed by GitHub).

**Comment by tyrcho** [comment]
Hi, any news ? we have stopped using hoarder until this is resolved ...

**Comment by danicheg** [comment]
Hi, @tyrcho did you find some ad-hoc workaround for preventing recompilation or maybe find how to replace 'hoarder'?

**Comment by tyrcho** [comment]
unfortunately not :(
